File: packages/vcr-kit/src/middleware.ts
import type { Request, Response, NextFunction, RequestHandler } from 'express';
import { verifyConsentReceipt } from './verifier.js';
import { VerifiableConsentReceipt, VerifyOptions } from './types.js';

export interface ConsentPresentOptions extends VerifyOptions {
  header?: string;
  extract?(req: Request): VerifiableConsentReceipt | undefined;
}

export function consentPresent(options: ConsentPresentOptions): RequestHandler {
  const header = options.header ?? 'x-consent-receipt';

  return async (req: Request, res: Response, next: NextFunction) => {
    try {
      const receipt = options.extract?.(req) ?? readReceiptFromRequest(req, header);
      if (!receipt) {
        res.status(403).json({ error: 'Consent receipt required' });
        return;
      }
      const result = await verifyConsentReceipt(receipt, options);
      if (!result.verified) {
        res.status(403).json({ error: result.reason ?? 'Consent receipt invalid' });
        return;
      }
      (req as Request & { consentReceipt?: VerifiableConsentReceipt }).consentReceipt = receipt;
      next();
    } catch (error: unknown) {
      next(error);
    }
  };
}

function readReceiptFromRequest(
  req: Request,
  headerName: string,
): VerifiableConsentReceipt | undefined {
  const headerValue = req.headers[headerName.toLowerCase()];
  if (typeof headerValue === 'string') {
    return parseReceipt(headerValue);
  }
  if (Array.isArray(headerValue) && headerValue.length > 0) {
    return parseReceipt(headerValue[0]);
  }
  const body = (req.body ?? {}) as Record<string, unknown>;
  const receipt = body.consentReceipt || body.consent_receipt;
  if (typeof receipt === 'string') {
    return parseReceipt(receipt);
  }
  if (typeof receipt === 'object' && receipt) {
    return receipt as VerifiableConsentReceipt;
  }
  return undefined;
}

function parseReceipt(value: string): VerifiableConsentReceipt | undefined {
  try {
    return JSON.parse(value) as VerifiableConsentReceipt;
  } catch (error) {
    return undefined;
  }
}

File: sdk/python/vcr_kit/middleware.py
"""ASGI middleware enforcing consent receipt presence."""

from __future__ import annotations

import json
from typing import Any, Awaitable, Callable, Dict

from .verifier import ConsentVerifier

ASGISend = Callable[[Dict[str, Any]], Awaitable[None]]
ASGIReceive = Callable[[], Awaitable[Dict[str, Any]]]
ASGIApp = Callable[[Dict[str, Any], ASGIReceive, ASGISend], Awaitable[None]]


class ConsentPresentMiddleware:
    """ASGI middleware that rejects requests without a valid consent receipt."""

    def __init__(
        self,
        app: ASGIApp,
        verifier: ConsentVerifier,
        *,
        header: str = "x-consent-receipt",
    ) -> None:
        self.app = app
        self.verifier = verifier
        self.header = header.lower()

    async def __call__(self, scope: Dict[str, Any], receive: ASGIReceive, send: ASGISend) -> None:
        if scope.get("type") != "http":
            await self.app(scope, receive, send)
            return

        header_value = _extract_header(scope, self.header)
        if not header_value:
            await _forbidden(send, "Consent receipt required")
            return

        try:
            receipt = json.loads(header_value)
        except json.JSONDecodeError:
            await _forbidden(send, "Consent receipt header must be JSON")
            return

        result = await self.verifier.verify(receipt)
        if not result.verified:
            await _forbidden(send, result.reason or "Consent receipt invalid")
            return

        scope.setdefault("state", {})["consent_receipt"] = receipt
        await self.app(scope, receive, send)


def consent_present(
    verifier: ConsentVerifier,
    *,
    header: str = "x-consent-receipt",
) -> Callable[[ASGIApp], ASGIApp]:
    """Wrap an ASGI app to enforce consent receipts."""

    def wrapper(app: ASGIApp) -> ASGIApp:
        return ConsentPresentMiddleware(app, verifier, header=header)

    return wrapper


def _extract_header(scope: Dict[str, Any], header: str) -> str | None:
    headers = scope.get("headers") or []
    for key, value in headers:
        if isinstance(key, bytes) and key.decode("latin-1").lower() == header:
            return value.decode("latin-1") if isinstance(value, bytes) else value
    return None


def _forbidden(send: ASGISend, message: str) -> Awaitable[None]:
    body = json.dumps({"error": message}).encode("utf-8")

    async def responder() -> None:
        await send(
            {
                "type": "http.response.start",
                "status": 403,
                "headers": [
                    (b"content-type", b"application/json"),
                    (b"content-length", str(len(body)).encode("ascii")),
                ],
            }
        )
        await send({"type": "http.response.body", "body": body})

    return responder()
